newzealidar/s3_connection.py

```diff
@@ -0,0 +1,217 @@
+        # Read the content of the retrieved object
+        body = resp["Body"].read()
+        # Determine the appropriate action based on the file extension of the object
+        # Check if the s3_object_key ends with ".pickle" extension
+        if s3_object_key.endswith(".pickle"):
+            # Deserialize (load) the binary data
+            data = pickle.loads(body)
+        # Check if the s3_object_key ends with ".nc" extension
+        elif s3_object_key.endswith(".nc"):
+            # Open the body as a BytesIO object for efficient in-memory handling
+            with BytesIO(body) as body_object:
+                # Load the dataset from the BytesIO object using the h5netcdf engine
+                data = xr.load_dataset(body_object, engine="h5netcdf")
+                # Check if the dataset does not have a Coordinate Reference System (CRS) defined
+                if data.rio.crs is None:
+                    try:
+                        # Extract and convert dataset's CRS spatial reference to EPSG code
+                        epsg_code = CRS.from_string(data.crs.spatial_ref).to_epsg()
+                    except AttributeError:
+                        # Extract and convert dataset's CRS spatial reference to EPSG code
+                        epsg_code = CRS.from_string(data.spatial_ref.crs_wkt).to_epsg()
+                    # Write the EPSG code as the Coordinate Reference System (CRS) for the dataset
+                    data.rio.write_crs(epsg_code, inplace=True)
+        # If the file extension is neither ".pickle" nor ".nc"
+        else:
+            # Read the content of the retrieved object using geopandas
+            data = gpd.read_file(BytesIO(body))
+        # Log a message confirming successful retrieval from the S3 bucket
+        log.info(f"Successfully retrieved '{s3_object_key}' from the S3 bucket.")
+        return data
+
+    def remove_object(self, s3_object_key: Union[str, pathlib.Path]) -> None:
+        """
+        Removes an object from the S3 bucket.
+        Parameters
+        ----------
+        s3_object_key : Union[str, pathlib.Path]
+            The key of the object to be removed from the S3 bucket. If a pathlib.Path object is provided,
+            it will be converted to a string representation.
+        """
+        # Check if the provided s3_object_key is a pathlib.Path object
+        if isinstance(s3_object_key, pathlib.Path):
+            # Convert the pathlib.Path object to a string representation
+            s3_object_key = s3_object_key.as_posix()
+        # Delete the object with the provided object key from the S3 bucket
+        self.s3_client.delete_object(Bucket=self.bucket_name, Key=s3_object_key)
+        # Log a message confirming successful deletion from the S3 bucket
+        log.info(f"Successfully deleted '{s3_object_key}' from the S3 bucket.")
+
+    def store_file(self, s3_object_key: Union[str, pathlib.Path], file_path: Union[str, pathlib.Path]) -> None:
```

Comment:
Could this take `file_path` as only `pathlib.Path` to help ensure people aren't putting incorrectly formatted `str` in here?

And do you think the same for `s3_object_key`? I'm not so familiar with the structure of them.

Reply:
I think this function should accept both `str` and `pathlib.Path` types. This is necessary because I think some parts of the codebase handle paths and directories as `strings` rather than converting them to `pathlib.Path` objects. Additionally, this approach ensures consistency with other methods in the S3Manager class.

Reply:
Keeping it the same for `s3_object_key`.

Reply:
Yes I see that it's normal for NewZealidar